### PR TITLE
added api for mcq count

### PR DESCRIPTION
**Pull request by Devyani606**

### CODE REVIEW

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 1]
```diff
@@ -82,18 +82,27 @@ def show
                          message: I18n.t('show.success', model_name: 'Assessment'))
         end
 
+        def count_hash_mcq # rubocop:disable Metrics/AbcSize
+          mcq_count_easy = Question.kept.tagged_with(['easy'], on: :difficulty).tagged_with(params[:mcq_tag_hash]).count
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:86]: Lets discuss this on call.

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 2]
```diff
@@ -32,15 +33,17 @@ def generate
...
+          @error_occured = false
+          validate_problems_and_mcq_count(@assessment.problem_count_hash, @assessment.mcq_count_hash)
           params[:email].split(',').each do |candidate_email|
             @drive = Drive.create!(name: @assessment.name, description: @assessment.description, organization_id: @organization.id,
-                                   is_assessment: true, uuid: @assessment.uuid, updated_by_id: User.kept.first.id,
-                                   created_by_id: User.kept.first.id, assessment_id: @assessment.id,
-                                   created_at: DateTime.current, updated_at: DateTime.current)
-            @assessment.problem_count.times { |problem| insert_problems(problem) }
-            find_candidate_send_email(candidate_email)
+                                   is_assessment: true, uuid: @assessment.uuid, updated_by_id: User.kept.first.id, created_by_id: User.kept.first.id,
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:40]: You can take in this in one variable `user_id  = User.kept.first.id` and use in `updated_by_id: user_id, created_by_id: user_id`

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 3]
```diff
@@ -78,16 +81,140 @@ def show
...
+        def mcq_count(difficulty)
+          if @assessment.mcq_tag_list.present?
+            Question.kept.tagged_with([difficulty], on: :difficulty).tagged_with(@assessment.mcq_tag_list, any: true).count
+          else
+            Question.kept.tagged_with([difficulty]).count
+          end
+        end
+
+        def total_count(problem_count_hash, mcq_count_hash)
+          problem_count_hash['easy'].to_i + problem_count_hash['medium'].to_i + problem_count_hash['hard'].to_i +
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:123]: If we have `problem_count_hash = {easy: 1, medium: 2, hard: 3}` then you can directly use `problem_count_hash.values.sum`
If `problem_count_hash = {easy: '1', medium: '2', hard: '3'}` then you can use `problem_count_hash.values.map(&:to_i).sum`

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 4]
```diff
@@ -78,16 +81,140 @@ def show
...
+        def validate_problems_and_mcq_count(problem_count_hash, mcq_count_hash)
+          return if total_count(problem_count_hash, mcq_count_hash).positive?
+
+          error_occured
+          render_error_message(message: I18n.t('count.required'))
+        end
+
+        def add_easy_problem(prob_count_hash)
+          if @easy_problem_count >= prob_count_hash['easy'].to_i
+            Problem.kept.tagged_with('easy').sample(prob_count_hash['easy'].to_i).each do |problem|
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:136]:
```
easy_count = prob_count_hash['easy'].to_i
@easy_problem_count >= easy_count
Problem.kept.tagged_with('easy').sample(easy_count).each do |problem|
```
Every time converting count hash data to integer you can store converted value in one variable and then use in other places.

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 5]
```diff
@@ -78,16 +81,140 @@ def show
...
+              DrivesProblem.create(drive_id: @drive.id, problem: problem)
+            end
+          else
+            error_occured
+            render_error_message(message: I18n.t('count.insufficient', model_name: 'problems', difficulty_level: 'Easy'))
+          end
+        end
+
+        def add_medium_problem(prob_count_hash)
+          if @medium_problem_count >= prob_count_hash['medium'].to_i
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:146]: same apply here

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 6]
```diff
@@ -78,16 +81,140 @@ def show
...
+            end
+          else
+            error_occured
+            render_error_message(message: I18n.t('count.insufficient', model_name: 'problems', difficulty_level: 'Medium'))
+          end
+        end
+
+        def add_hard_problem(prob_count_hash)
+          if @hard_problem_count >= prob_count_hash['hard'].to_i
+            Problem.kept.tagged_with('hard').sample(prob_count_hash['hard'].to_i).each do |problem|
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:158]: same here

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 7]
```diff
@@ -78,16 +81,140 @@ def show
...
+          return unless mcq_hash.present? || (mcq_hash['easy'].to_i +
+          mcq_hash['medium'].to_i + mcq_hash['hard'].to_i).positive?
+
+          add_easy_mcq(mcq_hash) unless @error_occured
+          add_medium_mcq(mcq_hash) unless @error_occured
+          add_hard_mcq(mcq_hash) unless @error_occured
+        end
+
+        def add_easy_mcq(mcq_hash)
+          if mcq_count('easy') >= mcq_hash['easy'].to_i
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:186]: same here for `mcq_hash['easy'].to_i`

File: db/migrate/20220127053222_add_problem_count_in_assessments.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+class AddProblemCountInAssessments < ActiveRecord::Migration[6.1]
+  def up
+    add_column :assessments, :problem_count_hash, :jsonb
+
+    Assessment.all.each do |assessment|
+      assessment.update(problem_count_hash: { easy: 1, medium: 1, hard: 1 })
+    end
+
+    remove_column :assessments, :problem_count
+  end
```
Comment (pooja-mane) [inline db/migrate/20220127053222_add_problem_count_in_assessments.rb:12]: need down method for this

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 9]
```diff
@@ -78,16 +81,140 @@ def show
...
+            end
+          else
+            error_occured
+            render_error_message(message: I18n.t('count.insufficient', model_name: 'problems', difficulty_level: 'Hard'))
+          end
+        end
+
+        def insert_problems(prob_count_hash)
+          return unless prob_count_hash.present? || (prob_count_hash['easy'].to_i +
+            prob_count_hash['medium'].to_i + prob_count_hash['hard'].to_i).positive?
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:169]: we need total count of problem and mcq also in separate methods like `insert_problems`, `insert_mcq` and `total_count`.
We can calculate both problem count and mcq count and store data in instance variable and use wherever we require.
If you want we can discuss this

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 10]
```diff
@@ -78,16 +81,140 @@ def show
...
+            render_error_message(message: I18n.t('count.insufficient', model_name: 'MCQs', difficulty_level: 'Easy'))
+          end
+        end
+
+        def add_medium_mcq(mcq_hash)
+          if mcq_count('medium') >= mcq_hash['medium'].to_i
+            Question.kept.tagged_with(['medium'], on: :difficulty).tagged_with(@assessment.mcq_tag_list,
+                                                                               any: true).sample(mcq_hash['medium'].to_i).each do |question|
+              DrivesQuestion.create(drive_id: @drive.id, question: question)
+            end
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:202]:
```
Question.kept.tagged_with(['medium'], on: :difficulty)
              .tagged_with(@assessment.mcq_tag_list, any: true)
              .sample(mcq_hash['medium'].to_i)
              .each do |question|
  DrivesQuestion.create(drive_id: @drive.id, question: question)
end
```
Indentation

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 11]
```diff
@@ -78,16 +81,140 @@ def show
...
+            render_error_message(message: I18n.t('count.insufficient', model_name: 'MCQs', difficulty_level: 'Medium'))
+          end
+        end
+
+        def add_hard_mcq(mcq_hash)
+          if mcq_count('hard') >= mcq_hash['hard'].to_i
+            Question.kept.tagged_with(['hard'], on: :difficulty).tagged_with(@assessment.mcq_tag_list,
+                                                                             any: true).sample(mcq_hash['hard'].to_i).each do |question|
+              DrivesQuestion.create(drive_id: @drive.id, question: question)
+            end
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:214]: Indentation

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 12]
```diff
@@ -78,16 +81,140 @@ def show
...
+          mcq_hash['medium'].to_i + mcq_hash['hard'].to_i).positive?
+
+          add_easy_mcq(mcq_hash) unless @error_occured
+          add_medium_mcq(mcq_hash) unless @error_occured
+          add_hard_mcq(mcq_hash) unless @error_occured
+        end
+
+        def add_easy_mcq(mcq_hash)
+          if mcq_count('easy') >= mcq_hash['easy'].to_i
+            Question.kept.tagged_with(['easy'], on: :difficulty).tagged_with(@assessment.mcq_tag_list,
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:187]: Indentation

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 2 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:40]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 4 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:136]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 5 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:146]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 6 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:158]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 7 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:186]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 3 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:123]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 9 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:169]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 10 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:202]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 11 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:214]: done

File: app/controllers/api/v1/admin/assessments_controller.rb (hunk 12 above)
Comment (Devyani606) [reply app/controllers/api/v1/admin/assessments_controller.rb:187]: done

File: app/controllers/api/v1/admin/assessments_controller.rb [hunk 13]
```diff
@@ -5,10 +5,10 @@ module V1
     module Admin
       class AssessmentsController < ApiController # rubocop:disable Metrics/ClassLength
         require 'will_paginate/array'
+        # before_action :authenticate_user!
```
Comment (pooja-mane) [inline app/controllers/api/v1/admin/assessments_controller.rb:8]: remove commented code

**Review by pooja-mane — APPROVED** [review]

**Pull request merged**